### PR TITLE
Rook 1.4.3 updates and Ekco 0.4.0

### PR DESCRIPTION
**Pull request by areed**
- Ekco 0.4.0 with no major changes besides image that will support Rook 1.4
- Enable shared fs in rook
- Do not redeploy ceph cluster when re-running kurl install script
- Start mon count at 1. Ekco will scale it up as nodes join.
- Change default osd pool size to 1 in global config map

### CODE REVIEW

File: addons/ekco/0.4.0/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+- `ekco-should-enable-purge-nodes`: Watch the cluster for dead nodes and remove them
+- `ekco-min-ready-master-node-count`: Don't purge the node if it will result in less than this many ready masters.
+   Only applicable if node purging is enabled.
+- `ekco-min-ready-worker-node-count`: Don't purge the node if it will result in less than this many ready workers
+   Only applicable if node purging is enabled.
+- `ekco-should-disable-clear-nodes`: Don't force delete pods on unreachable stuck in the terminating state
+- `ekco-disable-should-maintain-rook-storage-nodes`: Whether to maintain the list of nodes to use in the CephCluster config.
+   This also enables control of the replication factor of ceph pools, scaling up and down with the number of nodes in the cluster.
+- `ekco-disable-should-install-reboot-service`: Whether to disable the reboot service, responsible for graceful node termination.
+- `rook-version`: Version of Rook to manage
```
Comment (areed) [inline addons/ekco/0.4.0/README.md:18]: This is new, rest of file is unchanged from 0.3.0

File: addons/ekco/0.4.0/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          command: [ /usr/bin/ekco, operator ]
+          resources:
+            requests:
+              cpu: 100m
+              memory: 64Mi
+            limits:
+              cpu: 200m
+              memory: 128Mi
+          env:
+            - name: LOG_LEVEL
```
Comment (areed) [inline addons/ekco/0.4.0/deployment.yaml:33]: This is new, rest of file except image is unchanged from 0.3.0

File: addons/ekco/0.4.0/reboot/shutdown.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+while read -r uid; do
+        pod=$(kubectl get pods --all-namespaces -ojsonpath='{ range .items[*]}{.metadata.name}{"\t"}{.metadata.uid}{"\t"}{.metadata.namespace}{"\n"}{end}' | grep $uid )
+        kubectl delete pod $(echo $pod | awk '{ print $1 }') --namespace=$(echo $pod | awk '{ print $3 }') --wait=false
+done < <(lsblk | grep '^rbd[0-9]' | awk '{ print $7 }' | awk -F '/' '{ print $6 }')
+
+# delete local pods using the Ceph filesystem
+while read -r uid; do
+        pod=$(kubectl get pods --all-namespaces -ojsonpath='{ range .items[*]}{.metadata.name}{"\t"}{.metadata.uid}{"\t"}{.metadata.namespace}{"\n"}{end}' | grep $uid )
+        kubectl delete pod $(echo $pod | awk '{ print $1 }') --namespace=$(echo $pod | awk '{ print $3 }') --wait=false
+done < <(cat /proc/mounts | grep ':6789:/' | grep -v globalmount | awk '{ print $2 }' | awk -F '/' '{ print $6 }')
```
Comment (areed) [inline addons/ekco/0.4.0/reboot/shutdown.sh:17]: The `| grep -v globalmount` is the only thing new in here. Any node that has a pod mounting the shared fs now has an extra mount:

```
areed@areed-aka-21rl:~/kurl$ cat /proc/mounts | grep 6789
10.96.0.199:6789,10.96.2.45:6789,10.96.1.230:6789:/volumes/csi/csi-vol-cf576638-f93e-11ea-a492-8626cbbad91b/a5c1309b-8773-4d56-906f-8c4deda3f8ea /var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-a4829b95-6bec-4d58-a7fa-4b76378eb676/globalmount ceph rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=rook-shared-fs 0 0
10.96.0.199:6789,10.96.2.45:6789,10.96.1.230:6789:/volumes/csi/csi-vol-cf576638-f93e-11ea-a492-8626cbbad91b/a5c1309b-8773-4d56-906f-8c4deda3f8ea /var/lib/kubelet/pods/6341d404-91b8-4268-893f-aec1cf2c114d/volumes/kubernetes.io~csi/pvc-a4829b95-6bec-4d58-a7fa-4b76378eb676/mount ceph rw,relatime,name=csi-cephfs-node,secret=<hidden>,acl,mds_namespace=rook-shared-fs 0 0
```

File: addons/ekco/0.4.0/tmpl-configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    min_ready_worker_nodes: $EKCO_MIN_READY_WORKER_NODE_COUNT
+    maintain_rook_storage_nodes: $EKCO_SHOULD_MAINTAIN_ROOK_STORAGE_NODES
+    ceph_block_pool: replicapool
+    ceph_filesystem: rook-shared-fs
+    ceph_object_store: rook-ceph-store
+    min_ceph_pool_replication: 1
+    max_ceph_pool_replication: 3
+    certificates_dir: /etc/kubernetes/pki
+    reconcile_interval: 1m
+    rook_version: $ROOK_VERSION
```
Comment (areed) [inline addons/ekco/0.4.0/tmpl-configmap.yaml:20]: This is new, rest of file is unchanged.

File: addons/rook/1.4.3/cluster/tmpl-ceph-storage-class.yaml [hunk 5]
```diff
@@ -23,7 +23,7 @@ parameters:
   csi.storage.k8s.io/node-stage-secret-name: rook-csi-rbd-node
   csi.storage.k8s.io/node-stage-secret-namespace: rook-ceph
   # Specify the filesystem type of the volume. If not specified, csi-provisioner
-  # will set default as `ext4`. Note that `xfs` is not recommended due to potential deadlock
+  # will set default as ext4. Note that xfs is not recommended due to potential deadlock
```
Comment (areed) [inline addons/rook/1.4.3/cluster/tmpl-ceph-storage-class.yaml:26]: These were getting executed by bash

File: addons/rook/1.4.3/operator/config-override.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: rook-config-override
+  namespace: rook-ceph
+data:
+  config: |
+    [global]
+    osd pool default size = 1
```
Comment (areed) [inline addons/rook/1.4.3/operator/config-override.yaml:9]: default starts at 3, which causes a warning on singe node clusters. Ekco manages all pool sizes so the default is not important.

**Review by paha — APPROVED** [review]
LGTM

One note, I would rather have users explicitly enable `shared fs` if they need it then set it by default.

**Comment by areed** [comment]
I'll make a ticket for enabling shared fs

**Pull request merged**